src/BCFSerialize.cc

```diff
@@ -233,41 +300,62 @@ Status BCFReader::read(shared_ptr<bcf1_t>& ans) {
     }
     if ((size_t)current_ >= bufsz_)
         return Status::NotFound();
-    // FIXME bcf_raw_read_from_mem does not seem to check bounds -- potential security issue
-    int reclen = bcf_raw_read_from_mem(&buf_[current_], ans.get());
+    int reclen = bcf_raw_read_from_mem(buf_, current_, bufsz_, ans.get());
     current_ += reclen;
     return Status::OK();
 }
 
-/* Adapted from [htslib::vcf.c::bcf_hdr_read] to read
-   from memory instead of disk.
-*/
 Status BCFReader::read_header(const char* buf, int hdrlen, int& consumed, shared_ptr<bcf_hdr_t>& ans) {
-    if (strncmp(buf, "BCF\2\2", 5) != 0) {
-        return Status::Invalid("BCFReader::read_header invalid BCF2 magic string");
-    }
+    return bcf_raw_read_header(buf, hdrlen, consumed, ans);
+}
 
-    ans = shared_ptr<bcf_hdr_t>(bcf_hdr_init("r"), &bcf_hdr_destroy);
-    int loc = 5;
-    int hlen;
-    memcpy(&hlen, &buf[loc], 4);
-    loc += 4;
+// BCFScanner
+// constructor
+BCFScanner::BCFScanner(const char* buf, size_t bufsz) :
+    buf_(buf), bufsz_(bufsz)
+{}
 
-    // make sure we do not read beyond the buffer end
-    if (loc + hlen > hdrlen) {
-        return Status::Invalid("BCFReader::read_header truncated header");
-    }
+Status BCFScanner::Open(const char* buf,
+                       size_t bufsz,
+                       unique_ptr<BCFScanner>& ans) {
+    ans.reset(new BCFScanner(buf, bufsz));
+    return Status::OK();
+}
 
-    auto htxt = make_unique<char[]>(hlen);
-    memcpy(htxt.get(), &buf[loc], hlen);
+BCFScanner::~BCFScanner() {
+    buf_ = nullptr;
+    bufsz_ = 0;
+    current_ = 0;
+}
 
-    // FIXME bcf_hdr_parse does not seem to check bounds -- potential security issue
-    if (bcf_hdr_parse(ans.get(), htxt.get()) != 0) {
-        return Status::Invalid("BCFReader::read_header parse error");
-    }
+ // move the cursor to the next record
+Status BCFScanner::next() {
+    if ((size_t)current_ >= bufsz_)
+        return Status::NotFound();
+    current_ += bcf_raw_calc_rec_len(buf_, current_, bufsz_);
+    if ((size_t)current_ >= bufsz_)
+        return Status::NotFound();
+    return Status::OK();
+}
 
-    consumed = loc + hlen;
+Status BCFScanner::read(shared_ptr<bcf1_t>& ans) {
+    if (!ans) {
+        ans = shared_ptr<bcf1_t>(bcf_init(), &bcf_destroy);
+    }
+    bcf_raw_read_from_mem(buf_, current_, bufsz_, ans.get());
     return Status::OK();
 }
 
+// check if the current record overlaps a range
+bool BCFScanner::overlaps(const range &rng) {
+    return bcf_raw_overlap(buf_, current_, bufsz_, rng);
+}
+
+/* Adapted from [htslib::vcf.c::bcf_hdr_read] to read
+   from memory instead of disk.
+*/
+Status BCFScanner::read_header(const char* buf, int hdrlen, int& consumed, shared_ptr<bcf_hdr_t>& ans) {
```

Comment:
The coveralls report suggests we could change uses of `BCFReader::read_header` to `BCFScanner::read_header` and then get rid of `BCFReader` -- what do you think?